Update the advanced features documentation (#1591)
dottxt-ai/outlines: docs_new/features/advanced/logits_processors.md

---
title: Logits Processors
---

# Logits Processors

Logits processors are objects that control text generation by modifying the probability distribution of possible next tokens. They do this by adjusting the logits (raw model outputs) at each generation step, effectively biasing the model's token selection.
Processors can be used to:
1. Generate structured output (e.g., JSON that follows a specific schema)
2. Prevent the model from generating specific words or tokens
3. Implement custom token sampling strategies

## Overview

Outlines uses logits processors with steerable models—models that run locally and allow fine-grained control over the generation process. When using such models in Outlines, the output type provided is turned into a logits processor that is then passed to the inference engine.

> There are four models for which we rely on logits processors:

Reviewer: ```suggestion
There are four models that support logits processors:
```

- LlamaCpp
- MLXLM
- Transformers
- VLLMOffline

Instead of providing an output type that will be turned into a logits processor, it is possible to directly provide a logits processor. To do so, you must create a `Generator` instance using the `processor` keyword argument. You cannot directly call the model with a logits processor.

For instance:

```python
from outlines import Generator

model = ...
logits_processor = ...

# Create a generator with a custom logits processor
generator = Generator(model, processor=logits_processor)

> response = generator("Hello world")

Reviewer: Could we add a specific example here?
